File: increasingring/src/main/java/com/baybaka/increasingring/contoller/Controller.java
package com.baybaka.increasingring.contoller;

import android.media.AudioManager;

import com.baybaka.increasingring.RunTimeSettings;
import com.baybaka.increasingring.SettingsService;
import com.baybaka.increasingring.config.CachingConfigFactory;
import com.baybaka.increasingring.config.RingerConfig;
import com.baybaka.increasingring.receivers.TestPageOnCallEvenReceiver;
import com.baybaka.increasingring.service.RingToneT;
import com.baybaka.increasingring.utils.AudioManagerWrapper;
import com.baybaka.notificationlib.NotificationController;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Date;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

public class Controller {

    private final Logger LOG = LoggerFactory.getLogger(Controller.class.getSimpleName());

    private final SettingsService mSettingsService;
    private final CachingConfigFactory configFactory;
    private final RingToneT ringtone;

    private RunTimeSettings mRunTimeSettings;
    private final AudioManagerWrapper mAudioManagerWrapper;


    private volatile VolumeIncreaseThread currentThread;


    private boolean fastModeEnabled = false;
    private boolean ringWhenMute;

    private SoundRestorer mSoundRestorer;
    private NotificationController notificationController;

    private String lastNumber = "";
    private int numCount;

    private int numMaxCount;
    private long lastCallTimeForTHISNumber = new Date().getTime();
    private boolean findPhoneEnabled;

    @Inject
    public Controller(SettingsService settingsService, RunTimeSettings runTimeSettings,
                      AudioManagerWrapper audioManagerWrapper,
                      CachingConfigFactory configFactory, SoundRestorer soundRestorer,
                      NotificationController notificationController) {

        mSettingsService = settingsService;
        mRunTimeSettings = runTimeSettings;
        mAudioManagerWrapper = audioManagerWrapper;

        this.configFactory = configFactory;
        mSoundRestorer = soundRestorer;
        this.notificationController = notificationController;
        this.ringtone = new RingToneT(runTimeSettings.getContext());
        updateAllConfigs();
    }

    public void startVolumeIncrease(String callerNumber) {
        //to make sure no old increase task left
        stopVolumeIncrease();

        int ringerMode = mAudioManagerWrapper.getRingerMode();
        int preRingLevel = mAudioManagerWrapper.getCurrentChosenStreamVolume();
        if (mRunTimeSettings.isLoggingEnabled()) {
            LOG.debug("Inside increaseVolume, Ringer_MODE is {} and volume is {}", ringerMode, preRingLevel);
        }

        if (ringWhenMute || isPhoneInNormalRingMode()) {
            mSoundRestorer.saveCurrentSoundLevels();

            asapAction();
            checkForConfigUpdate();

            RingerConfig config = configFactory.getConfig(preRingLevel);
            currentThread = createThread(config, callerNumber) ; //new VolumeIncreaseThread(config, mAudioManagerWrapper, mRunTimeSettings, callerNumber);
            new Thread(currentThread).start();

            TestPageOnCallEvenReceiver.sendBroadcastToLogReceiver(mRunTimeSettings.getContext());

        } else {
            LOG.info("Abort increasing volume. Phone mode is {} and ringWhenMute function set to {}", ringerMode, false);
        }

        notificationController.startNotify();
    }

    /**
     * Fix for first call with max sound on some phones
     */
    private void asapAction() {
        if (fastModeEnabled) {
            mAudioManagerWrapper.setAudioLevelRespectingLogging(1);
        }
    }

    public void restoreVolumeToPreRingingLevel() {
        stopVolumeIncrease();
        mSoundRestorer.restoreVolumeToPreRingingLevel();
    }

    private void checkForConfigUpdate() {
        if (mRunTimeSettings.isConfigChanged()) {
            updateAllConfigs();
            mRunTimeSettings.configIsUpdated();
        }
    }

    public void updateAllConfigs() {
        mAudioManagerWrapper.changeOutputStream(getSoundStream());
        configFactory.updateStream(getSoundStream());

        configFactory.setMaxHardwareVolumeLevel(mAudioManagerWrapper.getChosenStreamrMaxHardwareVolumeLevel());
        configFactory.updateConfig();
        mSoundRestorer.updateConfig();
        updateLocal();
    }

    private void updateLocal() {
//        fastModeEnabled = settings.getAsapState();
        fastModeEnabled = true;
        ringWhenMute = mSettingsService.ringWhenMute();

        numMaxCount = mSettingsService.getFindPhoneCount();
        findPhoneEnabled = mSettingsService.isFindPhoneEnabled();
    }

    private boolean isPhoneInNormalRingMode() {
        return mAudioManagerWrapper.getRingerMode() == AudioManager.RINGER_MODE_NORMAL;
    }


    private int getSoundStream() {
        return mSettingsService.getSoundStream();
    }

    public void stopVolumeIncrease() {
        if (currentThread != null) {
            currentThread.stop();
            currentThread = null;
        }
        notificationController.stopNotify();

    }

    private VolumeIncreaseThread createThread(RingerConfig config, String callerNumber){
        int stream = mAudioManagerWrapper.getCurrentChosenStreamVolume();
        if( stream == AudioManager.STREAM_MUSIC){
            return new RingTone(config, mAudioManagerWrapper, mRunTimeSettings);
        } else {
            return new Music(config, mAudioManagerWrapper, mRunTimeSettings, callerNumber, ringtone);
        }
    }

    private void findPhoneMaximizeVolume(String callerNumber) {
        stopVolumeIncrease();
        mAudioManagerWrapper.changeOutputStream(AudioManager.STREAM_MUSIC);
        RingerConfig config = configFactory.getFindPhoneConfig();
        currentThread = createThread(config, callerNumber); //new VolumeIncreaseThread(config, mAudioManagerWrapper, mRunTimeSettings, callerNumber);
        new Thread(currentThread).start();

        mRunTimeSettings.configurationChanged(); // to reset stream on next call

//        mAudioManagerWrapper.maxVolDisableMuteVibrate();
//        updateAllConfigs();
        mRunTimeSettings.findPhoneNotification();
    }

    public void findPhone(String incomingNumber) {
        if (!findPhoneEnabled) return;
        long newCallTime = new Date().getTime();

        long minutes = TimeUnit.MILLISECONDS.toMinutes(newCallTime - lastCallTimeForTHISNumber);
        boolean timeIsOk = minutes < 6;

        if (lastNumber.equals(incomingNumber) && timeIsOk) {

            numCount++;
            if (numCount >= numMaxCount) {
                LOG.info("Calling find my phone function for number {}", incomingNumber);
                findPhoneMaximizeVolume(incomingNumber);
            }

        } else {
            numCount = 1;
            lastNumber = incomingNumber;
        }
        lastCallTimeForTHISNumber = newCallTime;
    }
}

File: increasingring/src/main/java/com/baybaka/increasingring/contoller/Controller.kt
package com.baybaka.increasingring.contoller

import com.baybaka.increasingring.audio.RingMode
import com.baybaka.increasingring.config.ConfigProvider
import com.baybaka.increasingring.receivers.TestPageOnCallEvenReceiver
import com.baybaka.increasingring.settings.RunTimeSettings
import com.baybaka.increasingring.settings.SettingsService
import com.baybaka.notificationlib.NotificationController
import org.slf4j.LoggerFactory
import javax.inject.Inject


class Controller @Inject
constructor(private val mSettingsService: SettingsService,
            private val mRunTimeSettings: RunTimeSettings,
//            private val mAudioManagerWrapper: IAudioController,
//            private val configFactory: CachingConfigFactory,
            private val config: ConfigProvider,
            private val mSoundRestorer: SoundRestorer,
            private val notificationController: NotificationController) {

    companion object{
        private val LOG = LoggerFactory.getLogger(Controller::class.java.simpleName)
    }

    private val phoneFinder: PhoneFinder
    private var currentThread: IVolumeIncreaseThread? = null

    private var ringWhenMute: Boolean = false


    init {
        this.phoneFinder = PhoneFinderImpl(mSettingsService)
        updateAllConfigs()
    }

    fun startVolumeIncrease(callerNumber: String) {

        //to make sure no old increase task left
        stopVolumeIncrease()

        checkForConfigUpdate()

        val searching = phoneFinder.lostMe(callerNumber)
        if (searching) {
            //use find phone
            findPhoneMaximizeVolume(callerNumber)
        } else {
            //use usual style
            standardFlow(callerNumber)
        }

        notificationController.startNotify()
    }

    private fun standardFlow(callerNumber: String) {

        val ringerMode = config.ringerMode
        someLogging()

        if (ringWhenMute || ringerMode === RingMode.RINGER_MODE_NORMAL) {
            mSoundRestorer.saveCurrentSoundLevels()

            currentThread = config.createThread(callerNumber)
            Thread(currentThread).start()

            TestPageOnCallEvenReceiver.sendBroadcastToLogReceiver(mRunTimeSettings.context)

        } else {
            LOG.info("Abort increasing volume. Phone mode is {} and ringWhenMute function set to {}", ringerMode, false)
        }
    }
    private fun someLogging(){
        if (mRunTimeSettings.isLoggingEnabled) {
            LOG.debug("Inside increaseVolume, Ringer_MODE is {}", config.ringerMode)
            config.printDebug()
        }
    }


    fun restoreVolumeToPreRingingLevel() {
        stopVolumeIncrease()
        mSoundRestorer.restoreVolumeToPreRingingLevel()
    }

    private fun checkForConfigUpdate() {
        if (mRunTimeSettings.isConfigChanged) {
            updateAllConfigs()
            mRunTimeSettings.configIsUpdated()
        }
    }

    fun updateAllConfigs() {
//        mAudioManagerWrapper.changeOutputStream(soundStream)
//        configFactory.updateStream(mSettingsService.soundStream)

        //todo should it do this itself?
//        configFactory.setMaxHardwareVolumeLevel(mAudioManagerWrapper.chosenStreamrMaxHardwareVolumeLevel)

        //todo same
        config.updateConfig()

        mSoundRestorer.updateConfig()
        phoneFinder.update()

//        mAudioManagerWrapper.changeStrategy(configFactory.getConfig())
        updateLocal()
    }

    private fun updateLocal() {
        ringWhenMute = mSettingsService.ringWhenMute()

    }


    fun stopVolumeIncrease() {
        currentThread?.let {
            it.stop()
            currentThread = null
        }

        notificationController.stopNotify()
    }


    private fun findPhoneMaximizeVolume(callerNumber: String) {
        LOG.info("Calling find my phone function for number $callerNumber")
        mSoundRestorer.saveCurrentSoundLevels()

        currentThread = config.createFindPhoneThread(callerNumber)
        Thread(currentThread).start()

        mRunTimeSettings.getNotifyProvider().findPhoneCalled()
    }

}
